### PR TITLE
[9.x] Add support for lazy loading commands

### PR DESCRIPTION
**Pull request by paras-malhotra**
Currently, all commands in Laravel have to be instantiated to be registered in the console application. That could be a problem for commands with big dependency graphs as all the dependencies are loaded in memory. See https://github.com/laravel/ideas/issues/1399.

Symfony has had lazy loading support for commands since version 3.4. This PR extends Symfony's lazy loading to Laravel. To allow for lazy loading, Laravel package console commands and/or app console commands just need to set the static `$defaultName` property like so:

```php
/**
 * The default command name for lazy loading.
 *
 * @var string|null
*/
protected static $defaultName = 'command:name';
```

Since the PR changes the return type of `Application::resolve` (adds a null return type), I'm sending this to master instead of 8x. If this PR is accepted, I can send across another PR to lazy load the commands shipped with Laravel.

### CODE REVIEW

File: src/Illuminate/Console/ContainerCommandLoader.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        return $this->container->get($this->commandMap[$name]);
+    }
+
+    /**
+     * Checks if a command exists.
+     *
+     * @param  string  $name
+     * @return bool
+     */
+    public function has(string $name)
```
Comment (antonkomarev) [inline src/Illuminate/Console/ContainerCommandLoader.php:58]: I suppose that you need to remove string typehint, Laravel doesn't use it

Comment (paras-malhotra) [reply src/Illuminate/Console/ContainerCommandLoader.php:58]: Thanks for the feedback. I just want to keep the method signatures consistent with the contract for now.

**Comment by taylorotwell** [comment]
@paras-malhotra it seems like we could make the getDefaultName() method have a default implementation in our console command that gets $signature before the first white-space. That way it wouldn't typically need to be defined in commands? Is that a bad idea?

**Comment by paras-malhotra** [comment]
@taylorotwell the `signature` is a non-static variable and that's the real issue here (cannot be accessed from static methods). The `getDefaultName` is a static method and it's structured that way in Symfony so that there's no need to instantiate the commands to access the default name.

**Comment by taylorotwell** [comment]
Thanks. Does it make sense to add a defaultName to some of Laravel's own heavier commands?

**Comment by paras-malhotra** [comment]
Yes @taylorotwell, I'll send across a follow-up PR to add default names to the commands shipped with Laravel. That should be a nice performance boost.

**Comment by garygreen** [comment]
@paras-malhotra does this actually work? It still has to resolve all of the console commands in order to get the name:

```php
            $this->artisan = (new Artisan($this->app, $this->events, $this->app->version()))
                                    ->resolveCommands($this->commands)
                                    ->setContainerCommandLoader();
```

... then when `resolvedCommands` is called:

```php
    /**
     * Add a command, resolving through the application.
     *
     * @param  string  $command
     * @return \Symfony\Component\Console\Command\Command|null
     */
    public function resolve($command)
    {
        if (class_exists($command) && ($commandName = $command::getDefaultName())) {
            $this->commandMap[$commandName] = $command;

            return null;
        }

        return $this->add($this->laravel->make($command));
    }
```

It calls `$command::getDefaultName()` which then LOADS the class in memory.

**Comment by garygreen** [comment]
I just tested with a large class based class file and when making static calls to a class it does indeed load it in memory as can be observed with `get_declared_classes` and `memory_get_usage` shows it's using memory.

It's probably still way lower than actually resolving the classes from the container, but still room for optimization so the classes aren't loaded from disk at all. Though I'm not sure how this plays with OpCache it may have those classes in memory already.

**Pull request merged**